python/cudf/cudf/core/buffer.py

```diff
@@ -5,108 +5,206 @@
+    Any,
+    Dict,
+    Mapping,
+    Protocol,
+    Tuple,
+    Union,
+    runtime_checkable,
+)
 
 import numpy as np
 
 import rmm
 
 import cudf
 from cudf.core.abc import Serializable
+from cudf.utils.string import format_bytes
+
+
+@runtime_checkable
+class DeviceBufferLike(Protocol):
+    def __getitem__(self, key: Union[int, slice]) -> DeviceBufferLike:
+        """Create a new view of the buffer."""
+
+    @property
+    def size(self) -> int:
+        """Size of the buffer in bytes."""
+
+    @property
+    def nbytes(self) -> int:
+        """Size of the buffer in bytes."""
+
+    @property
+    def ptr(self) -> int:
+        """Device pointer to the start of the buffer."""
+
+    @property
+    def owner(self) -> Any:
+        """Object owning the memory of the buffer."""
+
+    @property
+    def __cuda_array_interface__(self) -> Mapping:
+        """Implementation of the CUDA Array Interface."""
+
+    def memoryview(self) -> memoryview:
+        """Read-only access to the buffer through host memory."""
+
+    def serialize(self) -> Tuple[dict, list]:
+        """Serialize the buffer into header and frames.
+
+        Notice, device data is **not** moved to host memory necessarily.
+
+        Returns
+        -------
+        Tuple[Dict, List]
+            The first element of the returned tuple is a dict containing any
+            serializable metadata required to reconstruct the object. The
+            second element is a list containing the device data buffers
+            or memoryviews of the object.
+        """
+
+    @classmethod
+    def deserialize(cls, header: dict, frames: list) -> DeviceBufferLike:
+        """Generate an buffer from a serialized representation.
+
+        Parameters
+        ----------
+        header : dict
+            The metadata required to reconstruct the object.
+        frames : list
+            The Buffers or memoryviews that the object should contain.
+
+        Returns
+        -------
+        DeviceBufferLike
+            A new object that implements DeviceBufferLike.
+        """
+
+
+def as_device_buffer_like(obj: Any) -> DeviceBufferLike:
+    """
+    Factory function to wrap `obj` in a DeviceBufferLike object.
+
+    If `obj` isn't device-buffer-like already, a new buffer that implements
+    DeviceBufferLike and points to the memory of `obj` is created. If `obj`
+    represents host memory, it is copied to a new `rmm.DeviceBuffer` device
+    allocation. Otherwise, the data of `obj` is **not** copied.
+
+    The returned Buffer keeps a reference to `obj` in order to retain the
+    lifetime of `obj`.
+
+    Raises ValueError if the data of `obj` isn't C-contiguous.
+
+    Parameters
+    ----------
+    obj : buffer-like or array-like
+        An object that exposes either device or host memory through
+        `__array_interface__`, `__cuda_array_interface__`, or the
+        buffer protocol. Only when `obj` represents host memory are
+        data copied.
```

Comment:
The last sentence could be phrased more clearly.
```suggestion
        buffer protocol. If `obj` represents host memory, data will
        be copied.
```